coderd/insights.go

```diff
@@ -43,3 +51,390 @@ func (api *API) deploymentDAUs(rw http.ResponseWriter, r *http.Request) {
+			StartTime:   startTime,
+			EndTime:     endTime,
+			TemplateIDs: seenTemplateIDs,
+			Users:       userLatencies,
+		},
+	}
+	httpapi.Write(ctx, rw, http.StatusOK, resp)
+}
+
+// @Summary Get insights about templates
+// @ID get-insights-about-templates
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Insights
+// @Success 200 {object} codersdk.TemplateInsightsResponse
+// @Router /insights/templates [get]
+func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if !api.Authorize(r, rbac.ActionRead, rbac.ResourceDeploymentValues) {
+		httpapi.Forbidden(rw)
+		return
+	}
+
+	p := httpapi.NewQueryParamParser().
+		Required("start_time").
+		Required("end_time")
+	vals := r.URL.Query()
+	var (
+		// The QueryParamParser does not preserve timezone, so we need
+		// to parse the time ourselves.
+		startTimeString = p.String(vals, "", "start_time")
+		endTimeString   = p.String(vals, "", "end_time")
+		intervalString  = p.String(vals, string(codersdk.InsightsReportIntervalNone), "interval")
+		templateIDs     = p.UUIDs(vals, []uuid.UUID{}, "template_ids")
+	)
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: p.Errors,
+		})
+		return
+	}
+
+	startTime, endTime, ok := parseInsightsStartAndEndTime(ctx, rw, startTimeString, endTimeString)
+	if !ok {
+		return
+	}
+	interval, ok := verifyInsightsInterval(ctx, rw, intervalString)
+	if !ok {
+		return
+	}
+
+	var usage database.GetTemplateInsightsRow
+	var dailyUsage []database.GetTemplateDailyInsightsRow
+	// Use a transaction to ensure that we get consistent data between
+	// the full and interval report.
+	err := api.Database.InTx(func(db database.Store) error {
+		var err error
+
+		if interval != codersdk.InsightsReportIntervalNone {
+			dailyUsage, err = db.GetTemplateDailyInsights(ctx, database.GetTemplateDailyInsightsParams{
+				StartTime:   startTime,
+				EndTime:     endTime,
+				TemplateIDs: templateIDs,
+			})
+			if err != nil {
+				return xerrors.Errorf("get template daily insights: %w", err)
+			}
+		}
+
+		usage, err = db.GetTemplateInsights(ctx, database.GetTemplateInsightsParams{
+			StartTime:   startTime,
+			EndTime:     endTime,
+			TemplateIDs: templateIDs,
+		})
+		if err != nil {
+			return xerrors.Errorf("get template insights: %w", err)
+		}
+
+		return nil
+	}, nil)
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	intervalReports := []codersdk.TemplateInsightsIntervalReport{}
+	for _, row := range dailyUsage {
+		intervalReports = append(intervalReports, codersdk.TemplateInsightsIntervalReport{
+			StartTime:   row.StartTime,
+			EndTime:     row.EndTime,
+			Interval:    interval,
+			TemplateIDs: row.TemplateIDs,
+			ActiveUsers: row.ActiveUsers,
+		})
+	}
+
+	resp := codersdk.TemplateInsightsResponse{
```

Comment:
Do you think we can extract it to a `convert...` function?

Reply:
Do you think it would be beneficial? I don't foresee we need the re-use, but perhaps you feel it would  help readability?

Reply:
Mainly readability, I admit I got confused when saw a set of predefined apps.